### DISCUSSION TITLE
[CSV Import] map asset and opposing account to source and destination field

### DISCUSSION THREAD

**Discussion by lausserl** [firefly-iii/firefly-iii#5191] [Firefly III]
Hi, first of all thank you so much for your great work, this is an amazing project!

I am just getting into using it and want to start importing csv files. While importing, I discovered a problem that may be specific to my bank, but from looking at other import configurations maybe this happened more often and people found solutions for it.

Unfortunately my bank only delivers csv files with the structure of a source account and a destination account. Those are defined clearly, with name, IBAN and BIC, but my own asset account can be on the source or destination side. So when I try to define the roles for the columns, it is impossible to assign the roles correctly for everything to be imported. 
To summarize it, I tried a whole lot of different things. I can say it works if I change the opposing and asset configuration and import everything 2 times. 

First of all, is there an option I am missing to check source and destination fields for asset and opposing accounts? I already checked if I can receive a different csv file from my bank and tried some other options, but unfortunately none of the ideas worked, so if I can not find anything else I probably just import it 2 times with different configurations.

Here a short summary of what I have tried:
- Using opposing account for source and destination fields. It takes only the second field for a lookup, which leads to it using the default account for both if the destination account is also the default one.
- Same behavior with defining both as asset account, although I did not really think that would work ;)
- Just defining the IBAN of asset/opposing account to eliminate the chance of it trying to find something with the name.

I hope I was able to explain everything clear enough, please just ask if I forgot about explaining something about this very specific situation. Also my first time posting on gitlab, so please excuse if I forgot about some common courtesies.

**Comment by JC5** [comment]
Generally speaking, the choice which one is source or destination is also determined by the amount. In other words: if the amount is negative or positive determines the direction.

If your bank provides something like this, there is little I can do:

```
12.34;bank_account;shop
12.34;shop;bank_account
```

Could you send me some (anonimized) examples and your config file please?

**Comment by lausserl** [comment]
Ok yes that makes sense, I have seen that a lot of work has been put into recognizing this when I was looking for solutions for my issue.

The bank provides something like you wrote, but at least with the correct mathematical sign. So in your example, if the first one is positive, for the second one it would be negative to create an equal transfer. So for my csv file, it is determined by if the amount is negative or positive AND on which side my own account is positioned.

I will try to create an example for you, but will probably only be able to do it tomorrow, thanks for your quick reply already.

**Comment by JC5** [comment]
I'm curious to see a few examples, with as little censoring as you dare. The reason is, the logs are very detailed, so I'm curious to see what the CSV importer is doing exactly.

**Comment by lausserl** [comment]
So I changed the files but it contains the same structureas mine. 

To create context, I have mostly 3 kinds of transaction, 
- when I use a debit card, then I only get text in the description and no IBANS/names, to import those correctly I have to create rules 
- when someone interacts with my account (I get money or they take money) or 
- when I send money from my account. 

The problem I think is when I send money, they put my asset account on the sender field, so the code does not recognize it. That is why I included two config files, you will see when importing one of them, the interactions will fail that have my asset account on the wrong place :(.

So for using them I only created the asset account with the IBAN mentioned in the csv. All IBANs are just randomly generated, nothing serious in there I think. 

[test_git.csv](https://github.com/firefly-iii/firefly-iii/files/7388276/test_git.csv)

Since I can not upload json files, here are the config files with a csv ending, just rename them and you should be able to use them.
[test2.csv](https://github.com/firefly-iii/firefly-iii/files/7388286/test2.csv)
[test1.csv](https://github.com/firefly-iii/firefly-iii/files/7388287/test1.csv)

**Comment by lausserl** [comment]
Hi there I surely do not want to bother you, since you already helped me so much with understanding, but do you maybe have any suggestion on how I could solve this or should I just import it two times?

**Reply by JC5** [reply]
I found some edge cases that will be part of the next release, but there are some inconsistencies in your CSV files that I cannot fix.

<img width="1618" alt="Screenshot 2021-10-24 at 17 42 48" src="https://user-images.githubusercontent.com/5889984/138601669-ff69e46f-2620-48de-954d-85c35089d8a2.png">


- In the blue transaction, you buy stuff but you are the receiver and the amount is negative.
- In the red transaction, you send money away (fits) butt he amount is also positive.

**Reply by lausserl** [reply]
Yes, I realized that apparently my bank always references the amount to what happens with my asset account. So for your example, the blue one I buy something so negative amount and in the red one I transfer money to also negative :(.

It is a really weird way of phrasing transactions, but really amazing that it works to import two times, once in each direction. Apparently your software recognizes that the amount it relative to the asset account either way, so it actually is better than sender/receiver nomenclature.

Still leads to me having to import it two times though, but if the duplicate recognition is working this is not too bad. If you have any suggestions on how to solve the double import problem I am happy to test around more, otherwise I see this as (answered)resolved. Either way, thank you for your help and again for developing this great tool.

**Reply by JC5** [reply]
It's a strange way of organizing transactions, but double importing should work correctly.

I think this is it, glad I could help!
